### PR TITLE
remove far plane from perspective projection

### PR DESCRIPTION
**Pull request by joonazan**
I switched to near to 1 and far to 0, which should make depth much more
accurate.

3d doesn't have a far plane anymore, as it should be much more accurate than
before even with infinite Z.

Infinite-z does not make sense for orthographic projections, because error must
be small regardless of distance there.

### CODE REVIEW

File: amethyst_rendy/src/pass/debug_lines.rs [hunk 1]
```diff
@@ -222,7 +222,7 @@ fn build_lines_pipeline<B: Backend>(
                     blend: Some(pso::BlendState::ALPHA),
                 }])
                 .with_depth_test(pso::DepthTest {
-                    fun: pso::Comparison::LessEqual,
```
Comment (joonazan) [inline amethyst_rendy/src/pass/debug_lines.rs]: I'm a bit worried about this one. The skybox had LessEqual for good reason, but I fail to see why it was like this here.

GreaterEqual would be useful if debuglines was somehow modified to sometimes draw lines at infinity.

**Comment by joonazan** [comment]
Please rerun CI for me. It failed to fetch the repository on nightly.

File: amethyst_rendy/src/pass/debug_lines.rs (hunk 1 above)
Comment (joonazan) [reply amethyst_rendy/src/pass/debug_lines.rs]: Changed it to GreaterEqual. That causes later lines to be drawn on top of earlier ones. It should also make them visible in a 2d game.

**Comment by joonazan** [comment]
Now fetching failed on stable...

**Comment by azriel91** [comment]
bors try

**Comment by azriel91** [comment]
Heya, looks like some `camera` tests need to be updated.

```rust
failures:

---- camera::tests::extract_perspective_values stdout ----
thread 'camera::tests::extract_perspective_values' panicked at 'assert_ulps_eq!(0.1, proj.near())

    left  = 0.1
    right = inf

', amethyst_rendy/src/camera.rs:961:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

---- camera::tests::orthographic_depth_usage stdout ----
thread 'camera::tests::orthographic_depth_usage' panicked at 'assert_abs_diff_eq!(projected_point[2] / projected_point[3], 0.0)

    left  = 1.0
    right = 0.0

', amethyst_rendy/src/camera.rs:1206:9

---- camera::tests::extract_orthographic_values stdout ----
thread 'camera::tests::extract_orthographic_values' panicked at 'assert_ulps_eq!(-5.0, proj.near())

    left  = -5.0
    right = 100.0

', amethyst_rendy/src/camera.rs:992:9

---- camera::tests::perspective_depth_usage stdout ----
thread 'camera::tests::perspective_depth_usage' panicked at 'assert_abs_diff_eq!(projected_point[2] / projected_point[3], 0.0)

    left  = 1.000001
    right = 0.0

', amethyst_rendy/src/camera.rs:1180:9

---- camera::tests::screen_to_world_2d stdout ----
thread 'camera::tests::screen_to_world_2d' panicked at 'assert_ulps_eq!(camera.projection().screen_to_world_point(center_screen, diagonal, &transform), Point3::new(0.0, 0.0, -0.1))

    left  = Point { coords: Matrix { data: [0.0, 0.0, -2000.0] } }
    right = Point { coords: Matrix { data: [0.0, 0.0, -0.1] } }

', amethyst_rendy/src/camera.rs:862:9

---- camera::tests::screen_to_world_3d stdout ----
thread 'camera::tests::screen_to_world_3d' panicked at 'assert_ulps_eq!(camera.projection().screen_to_world_point(center_screen, diagonal, &transform), Point3::new(0.0, 0.0, -0.1))

    left  = Point { coords: Matrix { data: [0.0, 0.0, -1.0000001] } }
    right = Point { coords: Matrix { data: [0.0, 0.0, -0.1] } }

', amethyst_rendy/src/camera.rs:819:9

---- camera::tests::perspective_orientation stdout ----
thread 'camera::tests::perspective_orientation' panicked at 'assertion failed: `(left < right)`
  left: `0.1`,
 right: `0.0`', amethyst_rendy/src/camera.rs:1132:9

failures:
    camera::tests::extract_orthographic_values
    camera::tests::extract_perspective_values
    camera::tests::orthographic_depth_usage
    camera::tests::perspective_depth_usage
    camera::tests::perspective_orientation
    camera::tests::screen_to_world_2d
    camera::tests::screen_to_world_3d
```

**Comment by joonazan** [comment]
Yeah, looks like I missed those. What command should I use to run all tests?

**Comment by joonazan** [comment]
Those failures are legit and point to a bug I've introduced. I haven't updated the getters.

I want to remove them in a future PR, so I'll do that now. It would be a waste of effort to first change them and then remove them.

**Comment by joonazan** [comment]
Now the tests fail because the depth of a thing on the near plane is 1.000001 instead of 1.0. Can we make the test more lenient? I don't think I can improve the accuracy because there's just a single division.

**Comment by azriel91** [comment]
I haven't yet had time to review, but is the test with that inaccuracy using `assert_ulps_eq!`? That normally adjusts for floating point errors. If it already is, I think there's a version of the macro that allows taking in `n` times inaccuracy, perhaps set that to 2x.

**Comment by joonazan** [comment]
Yes, I'm using that macro.

I found the root cause of the inaccuracy. `0.1 / (3 - 2.9)` is 1.000001 in single precision. I'll fix that test by choosing values that can be represented in binary.

**Comment by joonazan** [comment]
Still TODO:

- [ ] rewrite auto_fov
- [ ] check that screen_to_world is pixel perfect

**Comment by joonazan** [comment]
I noticed that auto_fov does not work for orthographic cameras. Maybe it should be extended to take an enum of Perspective or Orthographic?

Also, `Camera::perspective` (previously `Projection::perspective`) makes a projection matrix so that the vertical fov is the fov given as argument. So I think I should make a more general version of it that does not assume that vertical fov is what you want to specify.

**Comment by azriel91** [comment]
> I noticed that auto_fov does not work for orthographic cameras. Maybe it should be extended to take an enum of Perspective or Orthographic?

Perhaps; or store it in the `Camera`? The previous behaviour *seems* to ignore changing the camera if it's not `Perspective`, but my game which uses an `Orthographic` camera and `AutoFovSystem` does inherit changes, so I think I went wrong somewhere when tracing code.

Also, I tried reviewing, but the math went over my head -- roundabout way of saying I'd trust you to get it right more than myself in reviewing :sweat_smile:. @jaynus / @Frizi, do either of you have time to review?

**Comment by azriel91** [comment]
Heya, I tried out this rebased on `master`, and the `sprites_ordered` / `sprite_camera_follow` examples don't render sprites (these use an orthographic camera).

I haven't managed to figure out how to update them to make them work, and can't seem to understand the code (sorry! but am getting lost in the math / code).

```bash
cargo run --example sprite_camera_follow --features "vulkan"
cargo run --example sprites_ordered --features "vulkan"
cargo run --example rendy --features "vulkan animation gltf"
```

**Comment by joonazan** [comment]
I'll look into it tomorrow.
19.2.2020 11.03 ip. "Azriel Hoh" <notifications@github.com> kirjoitti:

**Comment by joonazan** [comment]
@azriel91 I was able to reproduce the issue. Investigating it now.

**Comment by joonazan** [comment]
@azriel91 I fixed it. I had forgotten to change sprites to render with a depth test that accepts fragments with depth greater than the buffer.

I still want to improve the screen -> world and vice versa tests before this is merged. I have done some research on the topic and I've learned that in Vulkan screen coordinates 0.0 and 1.0 are the edges of the screen.

Is this true in Amethyst as well? For example, is mouse position measured like that, or is (0, 0) the center of the first pixel?

If there is no consensus, I'll use the Vulkan scheme and document it in the world_to_screen functions documentation.

**Comment by azriel91** [comment]
Heya, thanks for that (of course it's a one line change :stuck_out_tongue:).

> Is this true in Amethyst as well? For example, is mouse position measured like that, or is (0, 0) the center of the first pixel?

Based on [the current `master`](https://github.com/amethyst/amethyst/blob/d64745079e9ded753e3a05200ea32f2b31b370ef/amethyst_rendy/src/camera.rs#L861), it looks like do follow Vulkan's convention for screen coordinates -- `(0.0, 0.0)` top left, `(1.0, 1.0)` bottom right.

Thanks again for putting in so much work.

**Comment by azriel91** [comment]
By the way, even though screen coordinates are `0.0, 0.0` on the top left, world coordinates are `0.0, 0.0` on the bottom left by current convention. The UI coordinate system was changed to match world coordinates instead of screen coordinates at some point.

**Comment by joonazan** [comment]
@Azriel that only makes sense in 2d. In 3d it completely depends on the
camera orientation.

I would propose making default 2d, screen and UI all behave the same. I
believe this could be done by mirroring the y axis in the orthographic
projection matrix.

What is the best place to discuss things like this?
21.2.2020 11.03 ip. "Azriel Hoh" <notifications@github.com> kirjoitti:

**Comment by azriel91** [comment]
Heya, sorry forgot to reply over the weekend. I'd recommend https://community.amethyst.rs/ for initial discussion (and visibility) -- though it may switch over to https://github.com/amethyst/rfcs/ to be immortalized.

Personally I don't mind either orientation, as long as we stick with something.

**Comment by joonazan** [comment]
I think this PR is ready now.

**Comment by azriel91** [comment]
@joonazan hey, am going be releasing a version of `amethyst` tomorrow, but am sorry that am unable to comprehend the changes in this PR (math / graphics are difficult for myself), so have chosen to defer this until a later version. Promise to try get it in the next time 🙇

**Comment by jojolepro** [comment]
Needs a review from someone with the math knowledge to approve. If no review is made, we will merge if testing shows no negative impact, I guess?

**Comment by jojolepro** [comment]
All tests passing,  fly_camera and renderable examples working, merging.

bors r+
Thanks!

**Comment by jojolepro** [comment]
oh actually it doesn't build with all features enabled
@joonazan

**Comment by joonazan** [comment]
@jojolepro I made a simple fix. Did not expect that so many places depend on camera.

**Comment by joonazan** [comment]
@jojolepro Can you fix the book if it requires more fixing? I was not able to build it. It complains about missing module systems.

**Comment by AlveLarsson** [comment]
status? we are prepping for 0.15.1

**Comment by joonazan** [comment]
@AlveLarsson Only the book needs changes. I was unable to build the book on my computer. Probably almost nothing changes. I made an attempt, but the CI timed out like it sometimes does, so it didn't get to compiling the book.

**Comment by azriel91** [comment]
@joonazan Heya, I'll fix up the book some time today / tomorrow. Just gotta check if I can push to your branch

**Comment by joonazan** [comment]
@azriel91 I gave you access to my fork.

**Comment by azriel91** [comment]
build failure is file system lock race.

bors try

**Pull request merged**